compiler/src/dotty/tools/backend/sjs/JUnitBootstrappers.scala

```diff
@@ -0,0 +1,341 @@
+        val reifiedAnnot = New(mydefn.TestAnnotType, noArgConstr, Nil)
+        New(mydefn.TestMetadataType, List(name, ignored, reifiedAnnot))
+      }
+      JavaSeqLiteral(metadata, TypeTree(mydefn.TestMetadataType))
+    })
+  }
+
+  private def genInvokeTest(owner: ClassSymbol, testClass: ClassSymbol, tests: List[Symbol])(
+      implicit ctx: Context, mydefn: MyDefinitions): DefDef = {
+
+    val sym = ctx.newSymbol(owner, Names.invokeTest, Synthetic | Method,
+        MethodType(List(Names.instance, Names.name), List(defn.ObjectType, defn.StringType), mydefn.FutureType)).entered
+
+    DefDef(sym, { (paramRefss: List[List[Tree]]) =>
+      val List(List(instanceParamRef, nameParamRef)) = paramRefss
+      val castInstanceSym = ctx.newSymbol(sym, Names.castInstance, Synthetic, testClass.typeRef, coord = owner.span)
+      Block(
+        ValDef(castInstanceSym, instanceParamRef.cast(testClass.typeRef)) :: Nil,
+        tests.foldRight[Tree] {
+          val tp = mydefn.NoSuchMethodExceptionType
+          val constr = tp.member(nme.CONSTRUCTOR).suchThat { c =>
+            c.info.paramInfoss.head.size == 1 &&
+            c.info.paramInfoss.head.head.isRef(defn.StringClass)
+          }.symbol.asTerm
+          Throw(New(tp, constr, nameParamRef :: Nil))
+        } { (test, next) =>
+          If(Literal(Constant(test.name.toString)).select(defn.Any_equals).appliedTo(nameParamRef),
+              genTestInvocation(testClass, test, ref(castInstanceSym)),
+              next)
+        }
+      )
+    })
+  }
+
+  private def genTestInvocation(testClass: ClassSymbol, testMethod: Symbol, instance: Tree)(
+      implicit ctx: Context, mydefn: MyDefinitions): Tree = {
+
+    val resultType = testMethod.info.resultType
+    if (resultType.isRef(defn.UnitClass)) {
+      val newSuccess = ref(mydefn.SuccessModule_apply).appliedTo(ref(defn.BoxedUnit_UNIT))
+      Block(
+          instance.select(testMethod).appliedToNone :: Nil,
+          ref(mydefn.FutureModule_successful).appliedTo(newSuccess)
+      )
+    } else if (resultType.isRef(mydefn.FutureClass)) {
+      instance.select(testMethod).appliedToNone
+    } else {
+      // We lie in the error message to not expose that we support async testing.
+      ctx.error("JUnit test must have Unit return type", testMethod.sourcePos)
+      EmptyTree
+    }
+  }
+
+  private def genNewInstance(owner: ClassSymbol, testClass: ClassSymbol)(implicit ctx: Context): DefDef = {
+    val sym = ctx.newSymbol(owner, Names.newInstance, Synthetic | Method,
+        MethodType(Nil, defn.ObjectType)).entered
+
+    DefDef(sym, New(testClass.typeRef, Nil))
+  }
+
+  private def castParam(param: Symbol, clazz: Symbol)(implicit ctx: Context): Tree =
+    ref(param).cast(clazz.typeRef)
+
+  private def annotatedMethods(owner: ClassSymbol, annot: Symbol)(implicit ctx: Context): List[Symbol] =
+    owner.info.decls.filter(m => m.is(Method) && m.hasAnnotation(annot))
```

Comment:
Using .decls means you'll miss inherited members, is this OK here ? Also something that should be tested I think :).

Reply:
See https://github.com/scala-js/scala-js/blob/v1.0.0-M7/test-suite/js/src/test/scala/org/scalajs/testsuite/junit/JUnitSubClassTest.scala for tests of inherited members. Again, they will come after #6365.